src/iterators.rs

```diff
@@ -163,19 +161,17 @@ impl<'a, T> Iterator for PreOrderTraversal<'a, T> {
     type Item = &'a Node<T>;
 
     fn next(&mut self) -> Option<&'a Node<T>> {
-        let id = self.data.pop_front();
-
-        if let Some(ref node_id) = id {
-            let node_ref = self.tree.get_unsafe(node_id);
-
-            // prepend child_ids
-            for child_id in node_ref.children().iter().rev() {
-                self.data.push_front(child_id.clone());
-            }
-
-            return Some(node_ref);
-        }
-        None
+        self.data
+            .pop_front()
+            .and_then(|node_id| self.tree.get(&node_id).ok())
```

Comment:
Yeah, I agree it'd be nice if this were the case.  However, at this point the library has made a promise that you can't try to get a `Node` from `Tree` A if it lives in `Tree` B (but happens to have the same `NodeId`).

I supposed it could just return `None` in that case, but you might not ever know that you're accidentally mixing up which `Node`s belong to which `Tree`.  Regardless though, that would be a pretty large breaking change, so we'd be looking at a bump up to 2.0 for something like that.

This part of the API is something that I've had mixed feelings about ever since I designed it, but at this point that ship has sailed I think.  There are several things about the API that might need addressing if we were to go down this path.

Reply:
Yes that's what I thought too. Thought you could still change it in a major release.